AcmTable enhancements

dhaiducek opened the pull request:
- Add ability to provide actionResolver function (this will greatly simplify migrating GRC logic over to AcmTable)
- ~Add ability to embed extraToolbar into existing toolbar~ (Update: With the toolbar redesign, I've removed this toggle)
- ~Style pagination according to placement~ (Update: Pagination positioning has been removed via this PR following design recommendations)

KevinFCormier reviewed (commented):
Hi @dhaiducek,

For the toolbar and pagination changes, I assume you intend to use them together so that the appearance of the table in GRC does not change.
![image](https://user-images.githubusercontent.com/42188127/116701752-430c4c00-a996-11eb-80d1-1ff72b186374.png)

App has a similar table with the context switcher that currently looks like this:
![image](https://user-images.githubusercontent.com/42188127/116701705-34259980-a996-11eb-9d50-18f1322e7547.png)

I believe every other table currently has pagination on top. I do agree that it looks cleaner the way you have it. I think we probably need some guidance from design on whether they think having the pagination in different places on different pages is OK.

For the `actionsResolver` change, I think that's a good idea. I was considering doing this as well, but I ended up doing it external to the table just because of the way it interacts with table grouping.

But I don't see a new test case or story. Do you have a working prototype for this? I think rather than just using `IActionsResolver` directly, we may need a new interface where the item itself is passed to the actions resolver. If you look at line 635, the default case uses rowId to index an array internal to AcmTable to get the item. For the grouped/expandable case, we use the key which is available via the props in `rowData`, which is what gets passed to an `IActionsResolver`. However, the fact that we are storing the key in the props is an implementation detail of AcmTable, and not one that implementers of an action resolver should depend on.

dhaiducek commented:
@KevinFCormier For pagination/extraToolbar, that's correct, though if every other squad uses pagination on top, then maybe we would want to change ours also and revert my pagination changes for consistency... Our current designs do indeed continue using pagination on the bottom. I'll check with design for feedback.

For the `actionResolver`, I manually changed these lines in my local packages and it was working. Our designs call for an expandable row, which I'm working to implement now, so I'll update if it still seems to be working.

KevinFCormier commented:
> For the `actionResolver`, I manually changed these lines in my local packages and it was working. Our designs call for an expandable row, which I'm working to implement now, so I'll update if it still seems to be working.

I don't fully follow this. Are you only intending to use the `disableActions` property to fully remove actions for certain rows? You could set that for rows that you provide via `groupSummaryFn` or `addSubRows` but not for regular items in the table, right?

KevinFCormier commented on src/AcmTable/AcmTable.tsx:
```diff
@@ -192,8 +193,10 @@ export interface AcmTableProps<T> {
     groupSummaryFn?: (items: T[]) => IRow
     tableActions?: IAcmTableAction[]
     rowActions?: IAcmRowAction<T>[]
+    rowActionResolver?: IActionsResolver
```
```suggestion
    rowActionResolver?: (item: T) => IActions
```
You would then need to wrap this function before passing to the `Table` in order to resolve the item to pass, but I think it's more in line with the rest of our `AcmTable` interface and more powerful.

dhaiducek replied:
For the resolver, currently we pass a resolver function to the Patternfly `actionResolver` that conditionally renders actions in the menu as enabled or disabled based on the user's access to resources and namespaces. It also renders action titles such as "Enable"/"Disable" based on the policy's current specification and renders a separator.

Sorry, I'm not following how this change is different/more powerful than using the `IActionsResolver`. This change looks nearly identical to the Patternfly definition: 
```
export type IActionsResolver = (rowData: IRowData, extraData: IExtraData) => (IAction | ISeparator)[];
```
https://github.com/patternfly/patternfly-react/blob/7ecba96a191798d01c7f9209a14a04f4a658d7bb/packages/react-table/src/components/Table/TableTypes.ts#L151

dhaiducek commented:
I've talked pagination over with the design team. They agree that pagination should be at the top for all of RHACM. So, I'll update this PR to revert the `paginationAtBottom` change, though the embedded extraToolbar may be helpful still? Maybe?

dhaiducek commented:
I've added a simple test and story to this.

I also noticed that in Firefox the table doesn't scroll horizontally, so I added an `overflow-x` styling, but I'm not sure if that'd break something else and I'm misunderstanding something about the table.

Also, here's our resolver code that I'm trying to migrate for reference in case it might help explain why I'm trying to add this change: https://github.com/open-cluster-management/grc-ui/blob/cb8d60da67e2575fbce5a8579e676baa219cc3fb/src-web/components/modules/GrcToggleModule.js#L115-L187

KevinFCormier commented:
> I've added a simple test and story to this.
> 
> I also noticed that in Firefox the table doesn't scroll horizontally, so I added an `overflow-x` styling, but I'm not sure if that'd break something else and I'm misunderstanding something about the table.
> 
> Also, here's our resolver code that I'm trying to migrate for reference in case it might help explain why I'm trying to add this change: https://github.com/open-cluster-management/grc-ui/blob/cb8d60da67e2575fbce5a8579e676baa219cc3fb/src-web/components/modules/GrcToggleModule.js#L115-L187

Tables that are too wide have long been an issue. I'm not sure on this one. @jamestalton Do you have thoughts on the horizontal scroll?

For the actionsResolver, I agree it's a great feature to have. But I think you are getting the interface wrong, because the code you linked to and even your story code have to examine the contents of the table to identify the source item. For example, you have:

```
if (rowData['last-name'].title.indexOf('a') >= 0) {
```

Notice that the identifier `last-name` does not appear in the column definition. PatternFly generated it from the header, meaning this value would change if we ever had translation.

```
{
        header: 'Last Name',
        sort: 'last_name',
        cell: 'last_name',
        search: 'last_name',
}
```

Inside the `AcmTable` implementation, we set `props.key` on the row data, so this would allow us to find the item from the rowData we are given and pass that to the user's `rowActionResolver` function. This way the resolver has the full original data and it's not tied to the way the data is rendered or to the internals of `AcmTable`.

`AcmTable` has type parameter `T`, and nearly every other property that is a function that is related to data in the table has parameters of type `T`.

In addition, the `rowActions` property has its own type for actions:
```
rowActions?: IAcmRowAction<T>[]
```

Ideally the resolver should return an array of that type, though it is currently lacking support for separators, but that should be easily added.

dhaiducek commented:
Okay--thanks for all the explanations, @KevinFCormier! I'll think more on this today to figure out how to have our function use the original data and then mesh that into AcmTable.

dhaiducek commented on src/AcmTable/AcmTable.tsx:
```diff
@@ -605,22 +611,46 @@ export function AcmTable<T>(props: AcmTableProps<T>) {
+    let actionsResolver: IActionsResolver | undefined
+    if (rowActionResolver) {
+        actionsResolver = (rowData: IRowData) => {
+            const tableItem = rowData.props?.key && sorted.find((tableItem) => tableItem.key === rowData.props.key)
```
Is this the expected/desired way to identify the target row?

dhaiducek reviewed (commented):
Okay, I think I've sorted out the actionResolver some more. It feels more in line with AcmTable logic, though I'll need verification/validation. So much validation. 😆 

I've also been working on the GRC tables to match, injecting metadata into row objects so that we don't have to rely on the raw query data, which feels better anyway.

In addition, I've removed the horizontal table scroll and extraToolbarEmbed prop to focus the PR. I've also added `isDisabled` and `addSeparator` properties to the IAcmRowAction so that a separator can be injected before an action.

dhaiducek commented on src/AcmTable/AcmTable.tsx:
```diff
@@ -605,22 +616,54 @@ export function AcmTable<T>(props: AcmTableProps<T>) {
+    let actionResolver: IActionsResolver | undefined
+    if (rowActionResolver) {
+        actionResolver = (rowData: IRowData) => {
+            let tableItem
+            if (groupFn || addSubRows) {
+                tableItem = rowData.props?.key && sorted.find((tableItem) => tableItem.key === rowData.props.key)
+            } else {
+                tableItem = paged[rowData.secretTableRowKeyId]
+            }
```
Okay. I borrowed logic from `onClick` and found a `secretTableRowKeyId` that seemed to match the `rowId` so that I could pass the tableItem from AcmTable instead of from Patternfly Table. Does that seem right?

KevinFCormier replied:
```suggestion
            let tableItem
            if (groupFn || addSubRows) {
                tableItem = rowData.props?.key && paged.find((tableItem) => tableItem.key === rowData.props.key)
            } else if ((extraData?.rowIndex ?? -1) >= 0) {
                tableItem = paged[extraData?.rowIndex ?? 0]
            }
```

Same as for static actions, we can check just in the `paged` array. For the non-grouped/expandable case, I'm not sure if we should rely on `secretTableRowKeyId`. `IExtraData` has a `rowIndex` property which seems to work, but it is optional and I'm not really sure why.

KevinFCormier commented on src/AcmTable/AcmTable.tsx:
```diff
@@ -605,22 +616,54 @@ export function AcmTable<T>(props: AcmTableProps<T>) {
+                onClick: (_event: React.MouseEvent, rowId: number, rowData: IRowData) => {
+                    if (groupFn || addSubRows) {
+                        const tableItem =
+                            rowData.props?.key && sorted.find((tableItem) => tableItem.key === rowData.props.key)
```
```suggestion
                            rowData.props?.key && paged.find((tableItem) => tableItem.key === rowData.props.key)
```

`paged` will be a shorter list to search in

KevinFCormier reviewed (commented):
Thanks, Dale! This is looking good. I've added a few suggestions to consider.

I'm also trying to wrap my head around whether we should be using useCallback or useMemo for any of the actions-related functions. But my brain can't handle that at the moment, so I'll come back to it Monday.

KevinFCormier commented:
> Thanks, Dale! This is looking good. I've added a few suggestions to consider.
> 
> I'm also trying to wrap my head around whether we should be using useCallback or useMemo for any of the actions-related functions. But my brain can't handle that at the moment, so I'll come back to it Monday.

We can always come back to optimizations later, of course. But have a look at the suggestions, please. And have a good weekend!
